pkg/resolver/client/ens/ens.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ens
+
+import (
+	"errors"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/ethersphere/bee/pkg/resolver/client"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// Address is the swarm bzz address.
+type Address = swarm.Address
+
+// Make sure Client implements the resolver.Client interface.
+var _ client.Interface = (*Client)(nil)
+
+type dialType func(string) (*ethclient.Client, error)
+type resolveType func(bind.ContractBackend, string) (string, error)
+
+// Client is a name resolution client that can connect to ENS via an
+// Ethereum endpoint.
+type Client struct {
+	mu        sync.RWMutex
+	Endpoint  string
+	ethCl     *ethclient.Client
+	dialFn    dialType
+	resolveFn resolveType
+}
+
+// Option is a function that applies an option to a Client.
+type Option func(*Client)
+
+// NewClient will return a new Client.
+func NewClient(opts ...Option) *Client {
+	c := &Client{
+		dialFn:    wrapDial,
+		resolveFn: wrapResolve,
+	}
+
+	// Apply all options to the Client.
+	for _, o := range opts {
+		o(c)
+	}
+
+	return c
+}
+
+// Connect implements the resolver.Client interface.
+func (c *Client) Connect(ep string) error {
+	if c.dialFn == nil {
+		return errors.New("no dial function implementation")
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
This clock can be moved bellow the dialFn call and its error check. Mutex is only protecting the Client fields, and dialFn can take arbitrary long time and it should not hold a lock.